File: sdk/python-can-gsusb/examples/03_simple_fd.py
import can

from common import get_parser
from gsusb import GsUsbBus


def main() -> None:
    parser = get_parser("Single-channel CAN FD example")
    parser.add_argument("--gsusb-channel", type=int, default=0, help="Logical gs_usb channel (default: 0)")
    args = parser.parse_args()

    print(f"Opening {args.selector} Channel {args.gsusb_channel} FD")
    print(f"Arb: {args.bitrate}bps, Data: {args.data_bitrate}bps")
    bus = GsUsbBus(
        channel=args.selector,
        gsusb_channel=args.gsusb_channel,
        bitrate=args.bitrate,
        fd=True,
        data_bitrate=args.data_bitrate,
        termination_enabled=not args.no_termination,
    )
    try:
        msg = can.Message(
            arbitration_id=0x1FFFFFF0,
            is_extended_id=True,
            is_fd=True,
            bitrate_switch=True,
            data=[0x11, 0x22, 0x33, 0x44, 0x55, 0x66, 0x77, 0x88] + [0x00] * 56,
        )
        print("Sending FD frame (64 bytes)...")
        bus.send(msg)
        print("Listening...")
        while True:
            rx = bus.recv(1.0)
            if rx is not None:
                print(f"Received: {rx}")
    except KeyboardInterrupt:
        print("\nStopping...")
    finally:
        bus.shutdown()


if __name__ == "__main__":
    main()

File: sdk/gsusb-rs/examples/03_simple_fd.rs
#[path = "common/mod.rs"]
mod common;

use common::{make_bus_options, make_payload, parse_common_options, print_message};
use gsusb_rs::{CanMessage, GsUsbBus, Result};
use std::env;
use std::time::Duration;

fn main() -> Result<()> {
    let args = env::args().skip(1).collect::<Vec<_>>();
    let options = parse_common_options(&args, "Single-channel CAN FD example")?;
    let tx_bus = GsUsbBus::new(make_bus_options(&options, options.tx_channel, true, None, None))?;
    let rx_bus = GsUsbBus::new(make_bus_options(&options, options.rx_channel, true, None, None))?;

    let message = CanMessage {
        arbitration_id: 0x1FFF_FFF0,
        is_extended_id: true,
        is_fd: true,
        bitrate_switch: true,
        data: make_payload(&[0x11, 0x22, 0x33, 0x44, 0x55, 0x66, 0x77, 0x88], 64, 0x00),
        ..CanMessage::default()
    };
    print_message("Sending", &message);
    tx_bus.send(&message)?;
    let received = rx_bus.recv(Duration::from_secs(2)).ok_or_else(|| gsusb_rs::GsUsbError("No FD frame received".to_string()))?;
    print_message("Received", &received);
    tx_bus.shutdown();
    rx_bus.shutdown();
    Ok(())
}